src/snapshot/src/lib.rs

```diff
@@ -158,432 +160,215 @@ impl Snapshot {
             .unwrap();
-        snapshot_mem[20] = 123;
 
-        #[cfg(target_arch = "aarch64")]
-        let expected_err = Error::Crc64(0x1960_4E6A_A13F_6615);
-        #[cfg(target_arch = "x86_64")]
-        let expected_err = Error::Crc64(0x103F_8F52_8F51_20B1);
-
-        let load_result: Result<(Test1, _), Error> =
-            Snapshot::load(&mut snapshot_mem.as_slice(), 38, vm);
-        assert_eq!(load_result.unwrap_err(), expected_err);
+        assert!(matches!(
+            snapshot.load_with_version_check::<_, u8>(&mut data.as_slice(), data.len()),
+            Err(Error::Crc64(_))
+        ));
     }
 
-    #[allow(non_upper_case_globals)]
-    #[allow(non_camel_case_types)]
-    #[allow(non_snake_case)]
     #[test]
-    fn test_kvm_bindings_struct() {
-        #[repr(C)]
-        #[derive(Debug, PartialEq, Eq, Versionize)]
-        pub struct kvm_pit_config {
-            pub flags: ::std::os::raw::c_uint,
-            pub pad: [::std::os::raw::c_uint; 15usize],
-        }
-
-        let state = kvm_pit_config {
-            flags: 123_456,
-            pad: [0; 15usize],
-        };
-
-        let vm = VersionMap::new();
-        let mut snapshot_mem = vec![0u8; 1024];
-        // Serialize as v1.
-        let mut snapshot = Snapshot::new(vm.clone(), 1);
+    fn test_bad_version() {
+        let mut data = vec![0u8; 100];
+
+        // We write a snapshot with version "v1.3.12"
+        let snapshot = Snapshot::new(Version::new(1, 3, 12));
+        snapshot.save(&mut data.as_mut_slice(), &42u8).unwrap();
+
+        // Different major versions should not work
+        let snapshot = Snapshot::new(Version::new(2, 3, 12));
+        assert!(matches!(
+            snapshot.load_with_version_check::<_, u8>(&mut data.as_slice(), data.len()),
+            Err(Error::InvalidFormatVersion(Version {
+                major: 1,
+                minor: 3,
+                patch: 12,
+                ..
+            }))
+        ));
+        let snapshot = Snapshot::new(Version::new(0, 3, 12));
+        assert!(matches!(
+            snapshot.load_with_version_check::<_, u8>(&mut data.as_slice(), data.len()),
+            Err(Error::InvalidFormatVersion(Version {
+                major: 1,
+                minor: 3,
+                patch: 12,
+                ..
+            }))
+        ));
+
+        // We can't support minor versions bigger than ours
+        let snapshot = Snapshot::new(Version::new(1, 2, 12));
+        assert!(matches!(
+            snapshot.load_with_version_check::<_, u8>(&mut data.as_slice(), data.len()),
+            Err(Error::InvalidFormatVersion(Version {
+                major: 1,
+                minor: 3,
+                patch: 12,
+                ..
+            }))
+        ));
+
+        // But we can support minor versions smaller or equeal to ours. We also support
+        // all patch versions within our supported major.minor version.
+        let snapshot = Snapshot::new(Version::new(1, 4, 12));
```

Comment:
And here it should be `1.2.12` as supported as it is less than `1.3.12`

Reply:
See last comment